### PR TITLE
buffer until all logs are available: eth_getLogs

### LINKED ISSUE #7064: RPC response abruptly interrupted by timeout json error.
**Description**
RPC call being interrupted by timeout.

**Steps to Reproduce**
1. call rpc with array of request instead of just one in the same call. It was reported by someone on gnosis node.

**Actual behavior**
returns the following:
```
{
    "jsonrpc": "2.0",
    "result": [
        {
            "address": "0x0214014f69b11a80462f52a123fcbfa830e3fa11",
            "blockHash": "0x97f81688901dc30f3dd88706b56aeffa1ca599672fc862f9f8dcd0daff00ef56",
            "blockNumber": "0x2051f3c",
            "data": "0x000000000000000000000000f45e6e13ce8eaf623dac30193a28419293e7fabf0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000064e318b52b0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000e8a898d19847b16bb65c826fa1d9a735be6b70ee00000000000000000000000076e12efa4d44c4f495e82fba0792b0bf1af91f4f00000000000000000000000000000000000000000000000000000000",
            "logIndex": "0x0",
            "removed": false,
            "topics": [
                "0x4c8a9c748e976c17c2eb2c2bc50da76eac9cd90ff529f0fe900e0c10a179f031",
                "0x0000000000000000000000000000000000000000000000000000000000000000",
                "0x5e96df6c7831235f39530c213de1ee20b2780f5415b600f0859a813dad1d5d8b"
            ],
            "transactionHash": "0x17a378752998c623aa6ffa0093f0c71b8b090017568ff02039dd20c90dc1b125",
            "transactionIndex": "0x1",
            "transactionLogIndex": "0x0"
        }{
            "jsonrpc": "2.0",
            "error": {
                "code": -32016,
                "message": "Request was canceled due to enabled timeout."
            },
            "id": null
        }
```

**Expected behavior**
either return the error json part only, or continue producing the proper resoponse.

On latest release 1.26.0

### PR DESCRIPTION
**Pull request by rjnrohit**
Fixes #7064 
buffer until all logs are available: eth_getLogs
## Changes

- _List the changes_

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

#### Notes on testing

_Optional. Remove if not applicable._

#### Requires documentation update

- [ ] Yes
- [ ] No

_If yes, link the PR to the docs update or the issue with the details labeled `docs`. Remove if not applicable._

#### Requires explanation in Release Notes

- [ ] Yes
- [ ] No

_If yes, fill in the details here. Remove if not applicable._

## Remarks

_Optional. Remove if not applicable._

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs [hunk 1]
```diff
@@ -644,9 +645,25 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
...
+                foreach (FilterLog log in filterLogs)
+                {
+                    logs.Add(log);
+                }
+            }
+
+            if (logs.Count > _rpcConfig.MaxLogsPerResponse)
+            {
+                return ResultWrapper<IEnumerable<FilterLog>>.Fail($"Too many logs requested. Max logs per response is {_rpcConfig.MaxLogsPerResponse}.", ErrorCodes.LimitExceeded);
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:664]:
```suggestion
                logs.Dispose();
                return ResultWrapper<IEnumerable<FilterLog>>.Fail($"Too many logs requested. Max logs per response is {_rpcConfig.MaxLogsPerResponse}.", ErrorCodes.LimitExceeded);
            }
```

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs [hunk 2]
```diff
@@ -644,9 +645,25 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
...
+
+            using (cancellationTokenSource)
+            {
+                foreach (FilterLog log in filterLogs)
+                {
+                    logs.Add(log);
+                }
+            }
+
+            if (logs.Count > _rpcConfig.MaxLogsPerResponse)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:661]: this needs to be checked in above loop. If not you would be potentially adding milions of items to that collection.

Comment (rjnrohit) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:661]: I'm doing same as reth has done: https://github.com/paradigmxyz/reth/blob/89e55c4830a90092bd3a3fb7f9155dfea9d2749b/crates/rpc/rpc/src/eth/filter.rs#L502

Comment (rjnrohit) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs]: added the changes!

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs [hunk 3]
```diff
@@ -610,9 +611,25 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
...
+
+            using (cancellationTokenSource)
+            {
+                foreach (FilterLog log in filterLogs)
+                {
+                    if (logs.Count > _rpcConfig.MaxLogsPerResponse)
+                    {
+                        logs.Dispose();
+                        return ResultWrapper<IEnumerable<FilterLog>>.Fail($"Too many logs requested. Max logs per response is {_rpcConfig.MaxLogsPerResponse}.", ErrorCodes.LimitExceeded);
+                    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:627]: minor: I think this should be after `logs.Add` to be precise

**Review by LukaszRozmej — APPROVED** [review]

File: src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs [hunk 4]
```diff
@@ -106,6 +106,12 @@ public interface IJsonRpcConfig : IConfig
     [ConfigItem(Description = "The max length of HTTP request body, in bytes.", DefaultValue = "30000000")]
     long? MaxRequestBodySize { get; set; }
 
+
+    [ConfigItem(
+        Description = "The max number of logs per response. For method `eth_getLogs`.",
+        DefaultValue = "20000")]
+    public int? MaxLogsPerResponse {get;set;}
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs]: Why `nullable`?

Comment (rjnrohit) [reply src/Nethermind/Nethermind.JsonRpc/IJsonRpcConfig.cs]: I just looked at `MaxRequestBodySize` variable

File: src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs [hunk 5]
```diff
@@ -610,9 +611,25 @@ public ResultWrapper<IEnumerable<FilterLog>> eth_getLogs(Filter filter)
...
 
-            return ResultWrapper<IEnumerable<FilterLog>>.Success(GetLogs(filterLogs, cancellationTokenSource));
+            ArrayPoolList<FilterLog> logs = new(_rpcConfig.MaxLogsPerResponse);
+
+            using (cancellationTokenSource)
+            {
+                foreach (FilterLog log in filterLogs)
+                {
+                    logs.Add(log);
+                    if (logs.Count > _rpcConfig.MaxLogsPerResponse)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:624]: if `0` should we ignore this check?

Comment (rjnrohit) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:624]: for the case if the `MaxLogsPerResponse` is `0` ?

Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Eth/EthRpcModule.cs:624]: yes, should we treat 0 we may treat it as no limit

**Review by LukaszRozmej — APPROVED** [review]

**Review by benaadams — APPROVED** [review]

**Pull request merged**